hedera-node/hedera-networkadmin-service-impl/src/main/java/com/hedera/node/app/service/networkadmin/impl/handlers/FreezeHandler.java

```diff
@@ -72,51 +110,96 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
-
                 // FREEZE_UPGRADE and TELEMETRY_UPGRADE require a valid start_time and valid update_file and
                 // file_hash values
-            case FREEZE_UPGRADE, TELEMETRY_UPGRADE -> {
-                verifyFreezeStartTimeIsInFuture(freezeTxn, txnId.transactionValidStart());
-
-                // from proto specs, it looks like update file not required for FREEZE_UPGRADE and TELEMETRY_UPGRADE
-                // but specs aren't very clear
-                // current code in FreezeTransitionLogic checks for the file in specialFiles
-                // so we will do the same
-                verifyUpdateFileAndHash(freezeTxn, specialFileStore);
-            }
+            case FREEZE_ONLY, FREEZE_UPGRADE, TELEMETRY_UPGRADE -> verifyFreezeStartTimeIsInFuture(
+                    freezeTxn, txValidStart);
 
-                // FREEZE_ABORT does not require any additional checks
-            case FREEZE_ABORT -> {
+                // PREPARE_UPGRADE, FREEZE_ABORT do not require any additional checks
+            default -> {
                 // do nothing
             }
         }
+    }
 
-        // no need to add any keys to the context because this transaction does not require any signatures
-        // it must be submitted by an account with superuser privileges, that is checked during ingest
+    public void handle(
+            @NonNull final TransactionBody txn,
+            @NonNull final NetworkAdminServiceConfig adminServiceConfig,
+            @NonNull final ReadableSpecialFileStore specialFileStore,
+            @NonNull final WritableFreezeStore freezeStore) {
+        requireNonNull(txn);
+        requireNonNull(adminServiceConfig);
+        requireNonNull(specialFileStore);
+        requireNonNull(freezeStore);
+        final FreezeTransactionBody freezeTxn = txn.freezeOrThrow();
+        final FileID updateFileNum =
+                freezeTxn.updateFile(); // only some freeze types require this, it may be null for others
+
+        validateSemantics(freezeTxn, specialFileStore, updateFileNum);
+
+        final FreezeUpgradeActions upgradeActions = new FreezeUpgradeActions(adminServiceConfig, freezeStore);
+        final Timestamp freezeStartTime = freezeTxn.startTime(); // may be null for some freeze types
+        final Instant freezeStartTimeInstant = freezeStartTime == null
+                ? null
+                : Instant.ofEpochSecond(freezeStartTime.seconds(), freezeStartTime.nanos());
+
+        switch (freezeTxn.freezeType()) {
+            case PREPARE_UPGRADE ->
+            // by the time we get here, we've already checked that updateFileNum is non-null in validateSemantics()
+            upgradeActions.extractSoftwareUpgrade(specialFileStore
+                    .get(requireNonNull(updateFileNum).fileNum())
+                    .orElseThrow(() -> new IllegalStateException("Update file not found")));
+                // TODO: call networkCtx.recordPreparedUpgrade(freezeTxn); (issue #6201)
```

Comment:
Does the record get written immediately, or when the upgrade file is successfully extracted?
Also, could, if desired, use the following syntax to make the issue tie clearer
```java
// @todo('Issue #6201') call networkCtx.recordPreparedUpgrade(freezeTxn);
```